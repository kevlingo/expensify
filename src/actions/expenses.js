import uuid from 'uuid';

// Action generators
export const addExpense = ({
  id = uuid(),
  description = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id,
    description,
    note,
    amount,
    createdAt
  }
});
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});
