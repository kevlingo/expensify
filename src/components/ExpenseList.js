import React from 'react';
import { connect } from 'react-redux';
import ExpenseItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

const ExpenseList = ({ expenses }) => (
  <div>
    <h1>Expense List</h1>
    {expenses.map(expense => (
      <ExpenseItem key={expense.id} {...expense} />
    ))}
  </div>
);

const mapStateToProps = ({ expenses, filters }) => ({
  expenses: selectedExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);
