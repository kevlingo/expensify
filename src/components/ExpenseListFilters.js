import React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setEndDate,
  setStartDate
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {
  state = {
    calandarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = calandarFocused => {
    this.setState(() => ({ calandarFocused }));
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => {
            this.props.dispatch(
              e.target.value === 'amount' ? sortByAmount() : sortByDate()
            );
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calandarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => ({
  filters
});
export default connect(mapStateToProps)(ExpenseListFilters);
