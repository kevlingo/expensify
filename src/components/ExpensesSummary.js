import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selecteExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
  return (
    <h1>
      Viewing {expenseCount} {expenseCount === 1 ? 'expense' : 'expenses'}{' '}
      totalling {numeral(expenseTotal / 100).format('$0,0.00')}
    </h1>
  );
};

const mapStateToProps = state => {
  const visisbleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visisbleExpenses.length,
    exensesTotal: selecteExpensesTotal(visisbleExpenses)
  };
};
export default connect(mapStateToProps)(ExpensesSummary);
