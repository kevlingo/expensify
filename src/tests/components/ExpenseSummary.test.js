import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenes';

test('should render the expense summary correctly with one expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expenseTotal={235} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render the expense summary correctly with more than one expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={23} expenseTotal={2354432} />
  );
  expect(wrapper).toMatchSnapshot();
});
