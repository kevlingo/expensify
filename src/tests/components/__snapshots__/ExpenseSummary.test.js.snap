// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the expense summary correctly with more than one expense 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;

exports[`should render the expense summary correctly with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $0.00
  </h1>
</div>
`;
